### PR TITLE
src: cpu: aarch64: matmul: add support for matmul on aarch64

### PR DESCRIPTION
**Pull request by cfRod**
# Description

This PR introduces support for [matmul](https://oneapi-src.github.io/oneDNN/dev_guide_matmul.html) operation for AArch64 with the help of
Compute Library for the Arm Architecture (ACL) routines. It provides improved performance on NLP workloads on AArch64 CPUs compared to the reference variant. This implementation falls in line with the same approach as was detailed in the original RFC [#795](https://github.com/oneapi-src/oneDNN/pull/795).

## Outline

The key changes are listed below:

- New `acl_matmul.cpp/.hpp` files added in a new matmul sub-directory `src/cpu/aarch64` directory;
- New implementation `acl_matmul_t` added in `src/cpu/matmul/cpu_matmul_list.cpp`;
- An `acl_matmul_utils.cpp/.hpp` was also added.

# Checklist

## Code-change submissions

- [X] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`)
pass locally?

- [X] Have you formatted the code using clang-format?

### New features

- [N/A] Have you added relevant tests?

- [X] Have you provided motivation for adding a new feature?

###Bug fixes

- [N/A] Have you added relevant regression tests?

- [N/A] Have you included information on how to reproduce the issue (either in a github issue or in this PR)?

### CODE REVIEW

**Comment by cfRod** [comment]
Adding @diaena and @nSircombe from the team.

File: src/cpu/aarch64/matmul/acl_matmul.hpp
Comment (igorsafo) [inline src/cpu/aarch64/matmul/acl_matmul.hpp:71]: Verbose prints `primitive_kind` in a separate column of a verbose line, so there is no point in adding it to implementation name.

File: src/cpu/aarch64/matmul/acl_matmul.hpp
Comment (igorsafo) [inline src/cpu/aarch64/matmul/acl_matmul.hpp:143]: primitives are immutable, this is required to keep them in primitive cache, only its resources can be modified during execution. Is it possible to allocate mutex as part of resources?

File: src/cpu/aarch64/matmul/acl_matmul.hpp
Comment (igorsafo) [inline src/cpu/aarch64/matmul/acl_matmul.hpp:28]: Please avoid using `using namespace` in headers, because in this case a namespace is propagated into all object files which use this header.

File: src/cpu/aarch64/matmul/acl_matmul.hpp
Comment (igorsafo) [inline src/cpu/aarch64/matmul/acl_matmul.hpp:81]: flag `oscale_runtime` means that output scales can be either provided at primitive creation time in attributes or at execution time as an additional argument to the primitive. However during execution alpha is taken from attributes:
```cpp
// Set alpha (output scaling)
amp.alpha = attr.output_scales_.scales_[0];
```
This can be fixed by:
- Option A: prohibit `oscale_runtime` and allow `oscale` only
- Option B: Add a dispatching mechanism to get alpha from either attributes or arguments based on the value in attributes. In case of `oscale_runtime` oscale value in attributes will be equal to `DNNL_RUNTIME_F32_VAL`, which means that the scales will be provided at execution time.

Comment (cfRod) [reply src/cpu/aarch64/matmul/acl_matmul.hpp:81]: Thanks. I will make this change to option A.

File: src/cpu/aarch64/matmul/acl_matmul.hpp
Comment (cfRod) [reply src/cpu/aarch64/matmul/acl_matmul.hpp:141]: Hi @igorsafo , from what I understand this feature was added recently here https://github.com/oneapi-src/oneDNN/pull/1131 as a way to provided thread-safe access to resource_t. I would keep this inline with what added previously and then address the your comment as a separate PR for all primitives integrated with ACL.

File: src/cpu/aarch64/matmul/acl_matmul.hpp
Comment (cfRod) [reply src/cpu/aarch64/matmul/acl_matmul.hpp]: Ok, added the change.

File: src/cpu/aarch64/matmul/acl_matmul.hpp
Comment (cfRod) [reply src/cpu/aarch64/matmul/acl_matmul.hpp]: Changed.

File: src/cpu/aarch64/matmul/acl_matmul.hpp
Comment (igorsafo) [reply src/cpu/aarch64/matmul/acl_matmul.hpp:141]: I see, thank you for the explanation. Resolving.
Just to make it clear, with the current approach multiple similar **independent** matmul primitives will share the same mutex (due to cache mechanism and shared primitive entry in cache) so their execution will be sequential. Once mutex is placed in resource_t, each matmul will work with its own mutex so multiple matmul primitives can be executed concurrently.

**Review by igorsafo — APPROVED** [review]

**Comment by igorsafo** [comment]
@cfRod The changes LGTM, thanks!

File: src/cpu/aarch64/matmul/acl_matmul.hpp
Comment (nSircombe) [reply src/cpu/aarch64/matmul/acl_matmul.hpp:141]: Thanks @igorsafo, I don't think this is something we've actually considered before. As @cfRod suggested, we should take a look.

**Comment by cfRod** [comment]
Hi @igorsafo , is it possible to get this into rls2.4? I have an identical PR ready here: https://github.com/oneapi-src/oneDNN/pull/1161.
Thanks for approving!

**Comment by igorsafo** [comment]
@cfRod Could you please take a look at SVE512 testing? I see that it was killed by timeout, so we need to reduce amount of tests or make sure they don't fall into the reference implementation

**Comment by nSircombe** [comment]
Hi @igorsafo, I think it might be a transient issue - I notice the CI on the previous patchset ran to completion in a little over 30': https://cloud.drone.io/oneapi-src/oneDNN/868/7/2

And the SVE512 case in the PR against the r2.4 branch has run successfully (in about 24'): https://cloud.drone.io/oneapi-src/oneDNN/874/7/2 (the other SVE stages are yet to complete at time of writing though).

The SVE512 stage doesn't build with `-DDNNL_AARCH64_USE_ACL=ON` set, so the changes in this PR _shouldn't_ impact the SVE stages of the DroneCI.

All the same, I did some manual testing of the PR, on A64fx (i.e. no QEMU!), and observe 100% of the tests passing in about 160s. Running some examples with `DNNL_VERBOSE=1` shows `dnnl_verbose,exec,cpu,convolution,jit:sve_512,...` as expected.

Maybe worth re-triggering the pipelines?
I don't see a way to do this in the DroneCI UI though - possibly a `git commit --amend --no-edit` is called for.

**Comment by cfRod** [comment]
ok thanks @igorsafo @nSircombe. I have a small update to my patch to catch errors from ACL's validate method in oneDNN (to ensure there is a fallback to reference kernels if ACL validate fails). I will update both PRs shortly. Apologies, I found this issue yesterday and thought I'd update this patch before it gets merged!

**Comment by cfRod** [comment]
Hi @igorsafo,
Is this waiting on the comment on the testing on the other PR?

**Comment by igorsafo** [comment]
Hi @cfRod ,
No, this PR will be merged by EoW.

**Comment by cfRod** [comment]
>No, this PR will be merged by EoW.

Ah, Thanks!

**Comment by dzarukin** [comment]
Thank you for contribution!

**Pull request merged**
